File: study/ts/script/utils/event/SlotList.ts
import { Slot } from "./Slot";

export class SlotList {
    public head: Slot;
    public tail: SlotList;
    public nonEmpty: boolean;

    public static NIL = new SlotList(null, null);

    constructor(head: Slot, tail?: SlotList) {
        if (head == null && tail == null) {
            this.nonEmpty = false;
        }
        else if (head == null) {

        }
        else {
            this.head = head;
            this.tail = tail || SlotList.NIL;
            this.nonEmpty = true;
        }
    }

    public getLength() {
        if (!this.nonEmpty) {
            return 0
        }
        if (this.tail == SlotList.NIL) {
            return 1
        }
        let result: number = 0;
        let p: SlotList = this;
        while (p.nonEmpty) {
            result = result + 1
            p = p.tail;
        }
        return result;
    }

    public prepend(slot) {
        return new SlotList(slot, this)
    }

    public append(slot) {
        if (!slot) {
            return this
        }
        if (!this.nonEmpty) {
            return new SlotList(slot)
        }

        if (this.tail == SlotList.NIL) {
            return new SlotList(slot).prepend(this.head)
        }

        let wholeClone = new SlotList(this.head)
        let subClone = wholeClone
        let current = this.tail

        while (current.nonEmpty) {
            let currentClone = new SlotList(current.head)
            subClone.tail = currentClone
            subClone = currentClone
            current = current.tail
        }
        subClone.tail = new SlotList(slot)
        return wholeClone
    }

    public insertWithPriority(slot: Slot) {
        if (!this.nonEmpty) {
            return new SlotList(slot)
        }

        let priority = slot.getPriority()
        if (priority > this.head.getPriority()) {
            return this.prepend(slot)
        }

        let wholeClone = new SlotList(this.head)
        let subClone = wholeClone
        let current = this.tail

        while (current.nonEmpty) {
            if (priority > current.head.getPriority()) {
                subClone.tail = current.prepend(slot)
                return wholeClone
            }
            let currentClone = new SlotList(current.head)
            subClone.tail = currentClone
            subClone = currentClone
            current = current.tail
        }

        subClone.tail = new SlotList(slot)
        return wholeClone
    }

    public filterNot(listener: Function) {
        if (!this.nonEmpty || listener == null) {
            return this
        }

        if (listener == this.head.getListener()) {
            return this.tail
        }

        let wholeClone = new SlotList(this.head)
        let subClone = wholeClone
        let current = this.tail
        while (current.nonEmpty) {
            if (current.head.getListener() == listener) {
                subClone.tail = current.tail
                return wholeClone
            }
            let currentClone = new SlotList(current.head)
            subClone.tail = currentClone
            subClone = currentClone
            current = current.tail
        }
        return this
    }

    public contains(listener) {
        if (!this.nonEmpty) {
            return false
        }
        let p: SlotList = this
        while (p.nonEmpty) {
            if (p.head.getListener() == listener) {
                return true
            }
            p = p.tail
        }

        return false
    }

    public find(listener) {
        if (!this.nonEmpty) {
            return null
        }

        let p: SlotList = this
        while (p.nonEmpty) {
            if (p.head.getListener() == listener) {
                return p.head
            }
            p = p.tail
        }
        return null
    }
}

File: study/lua/src/yoka/event/SlotList.lua
local SlotList = class("SlotList")

SlotList.NIL = SlotList.new(nil, nil)
--
function SlotList:ctor(head, tail)
	if not head and not tail then
		self.nonEmpty = false
	elseif not head then
	else
		self.head = head
		self.tail = tail or SlotList.NIL
		self.nonEmpty = true
	end
end

--
function SlotList:getLength()
	if not self.nonEmpty then return 0 end
	if self.tail == SlotList.NIL then return 1 end

	local result = 0
	local p = self
	while p.nonEmpty do
		result = result + 1
		p = p.tail
	end

	return result
end

--
function SlotList:prepend(slot)
	return SlotList.new(slot, self)
end

--
function SlotList:append(slot)
	if not slot then return self end
	if not self.nonEmpty then return SlotList.new(slot) end

	if self.tail == SlotList.NIL then
		return SlotList.new(slot):prepend(self.head)
	end

	local wholeClone = SlotList.new(self.head)
	local subClone = wholeClone
	local current = self.tail

	while current.nonEmpty do
		local currentClone = SlotList.new(current.head)
		subClone.tail = currentClone
		subClone = currentClone
		current = current.tail
	end

	subClone.tail = SlotList.new(slot)
	return wholeClone
end

--
function SlotList:insertWithPriority(slot)
	if not self.nonEmpty then return SlotList.new(slot) end

	local priority = slot:getPriority()
	if priority > self.head:getPriority() then return self:prepend(slot) end

	local wholeClone = SlotList.new(self.head)
	local subClone = wholeClone
	local current = self.tail

	while current.nonEmpty do
		if priority > current.head:getPriority() then
			subClone.tail = current:prepend(slot)
			return wholeClone
		end
		local currentClone = SlotList.new(current.head)
		subClone.tail = currentClone
		subClone = currentClone
		current = current.tail
	end

	subClone.tail = SlotList.new(slot)
	return wholeClone
end

--
function SlotList:filterNot(listener)
	if not self.nonEmpty or listener == nil then return self end

	if listener == self.head:getListener() then return self.tail end

	local wholeClone = SlotList.new(self.head)
	local subClone = wholeClone
	local current = self.tail

	while current.nonEmpty do
		if current.head:getListener() == listener then
			subClone.tail = current.tail
			return wholeClone
		end
		local currentClone = SlotList.new(current.head)
		subClone.tail = currentClone
		subClone = currentClone
		current = current.tail
	end

	return self
end

--
function SlotList:contains(listener)
	if not self.nonEmpty then return false end

	local p = self
	while p.nonEmpty do
		if p.head:getListener() == listener then return true end
		p = p.tail
	end

	return false
end

--
function SlotList:find(listener)
	if not self.nonEmpty then return nil end

	local p = self
	while p.nonEmpty do
		if p.head:getListener() == listener then return p.head end
		p = p.tail
	end

	return nil
end



return SlotList